### PR TITLE
Dt 694

### PR DESCRIPTION
**Pull request by JB-softwire**
Edit username and details endpoints.

### CODE REVIEW

File: auth-service/src/main/kotlin/uk/gov/communities/delta/auth/controllers/internal/EditAccessGroupsController.kt [hunk 1]
```diff
@@ -59,12 +123,67 @@ class EditAccessGroupsController(
...
+        groupName: String,
+        callingUser: LdapUser,
+        targetUser: LdapUser,
+        allAccessGroupNames: List<String>,
+    ) {
+        validateGroupExistenceAndUserAuthority(allAccessGroupNames, groupName, callingUser)
+
+        if (targetUser.memberOfCNs.contains(LDAPConfig.DATAMART_DELTA_PREFIX + groupName)) {
+            throw ApiError(
+                HttpStatusCode.Forbidden,
```
Comment (BenRamchandani) [inline auth-service/src/main/kotlin/uk/gov/communities/delta/auth/controllers/internal/EditAccessGroupsController.kt:141]: Nit most of these should be 400 (or maybe 409 or 422), providing different credentials is never going to make this request succeed.

**Review by BenRamchandani — COMMENTED** [review]
This looks good, I'll have another proper look next week

File: auth-service/src/main/kotlin/uk/gov/communities/delta/auth/controllers/internal/AdminEditEmailController.kt [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+import io.ktor.server.routing.*
+import kotlinx.serialization.SerialName
+import kotlinx.serialization.Serializable
+import org.slf4j.LoggerFactory
+import uk.gov.communities.delta.auth.plugins.ApiError
+import uk.gov.communities.delta.auth.services.DeltaSystemRole
+import uk.gov.communities.delta.auth.services.UserLookupService
+import uk.gov.communities.delta.auth.services.UserService
+
+class AdminEditEmailController(
```
Comment (BenRamchandani) [inline auth-service/src/main/kotlin/uk/gov/communities/delta/auth/controllers/internal/AdminEditEmailController.kt:16]: Maybe `AdminEditUserEmailController`? Just to make it clearer editing a user, not an email

File: auth-service/src/main/kotlin/uk/gov/communities/delta/auth/controllers/internal/AdminEditEmailController.kt [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+            arrayOf(DeltaSystemRole.ADMIN), call
+        )
+        val requestData = call.receive<DeltaEmailChangeRequest>()
+        val requestedEmail = requestData.newEmail
+
+        val userToEdit = userLookupService.lookupUserByCn(requestData.userToEditCn)
+
+        logger.atInfo().log("Updating email and username for user {}", userToEdit.cn)
+
+        validateEmail(requestedEmail)
```
Comment (BenRamchandani) [inline auth-service/src/main/kotlin/uk/gov/communities/delta/auth/controllers/internal/AdminEditEmailController.kt:38]: Let's log lots of detail

```suggestion
        validateEmail(requestedEmail)
        logger.atInfo().addKeyValue("oldUserEmail", userToEdit.email)
            .addKeyValue("oldUserCN", userToEdit.cn)
            .addKeyValue("newUserEmail", requestedEmail)
            .log("Request to update email and username for user")
```

File: auth-service/src/main/kotlin/uk/gov/communities/delta/auth/controllers/internal/AdminEditEmailController.kt [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+        userService.updateEmail(userToEdit, requestedEmail, session, call)
+        return call.respond(mapOf("message" to "Username has been updated."))
+    }
+
+    private fun validateEmail(requestedUsername: String) {
```
Comment (BenRamchandani) [inline auth-service/src/main/kotlin/uk/gov/communities/delta/auth/controllers/internal/AdminEditEmailController.kt:44]: Maybe we should check it's a valid email as well. I know we don't strictly require it, but in practice for editing users in the frontend I think it's fine to. There's an `EmailAddressChecker` you can use

File: auth-service/src/main/kotlin/uk/gov/communities/delta/auth/controllers/internal/EditAccessGroupsController.kt [hunk 5]
```diff
@@ -59,12 +123,67 @@ class EditAccessGroupsController(
...
         val accessGroupActions = generateAccessGroupActions(accessGroupRequestMap, currentAccessGroups, selectedOrgs)
 
-        executeAccessGroupActions(accessGroupActions, session, callingUser, call)
+        executeAccessGroupActions(accessGroupActions, null, callingUser, call)
 
         return call.respond(mapOf("message" to "Access groups have been updated. Any changes to your roles or access groups will take effect the next time you log in."))
     }
 
-    private suspend fun validateRequest(
+    fun validateAddAccessGroupRequest(
```
Comment (BenRamchandani) [inline auth-service/src/main/kotlin/uk/gov/communities/delta/auth/controllers/internal/EditAccessGroupsController.kt:131]: I imagine the grant wizard also adds the user to `dclg` in the access group, can you check?

File: auth-service/src/main/kotlin/uk/gov/communities/delta/auth/controllers/internal/EditAccessGroupsController.kt [hunk 6]
```diff
@@ -28,22 +29,85 @@ class EditAccessGroupsController(
...
+
+        val allAccessGroups = accessGroupsService.getAllAccessGroups()
+        validateAddAccessGroupRequest(targetGroupName, callingUser, targetUser, allAccessGroups.map { it.name })
+
+        groupService.addUserToGroup(
+            targetUser.cn,
+            targetUser.dn,
+            targetGroupADName,
+            call,
+            null,
```
Comment (BenRamchandani) [inline auth-service/src/main/kotlin/uk/gov/communities/delta/auth/controllers/internal/EditAccessGroupsController.kt:55]: We should pass in the session here and below

File: auth-service/src/main/kotlin/uk/gov/communities/delta/auth/controllers/internal/EditAccessGroupsController.kt (hunk 5 above)
Comment (JB-softwire) [reply auth-service/src/main/kotlin/uk/gov/communities/delta/auth/controllers/internal/EditAccessGroupsController.kt]: We determined that it does but can you explain what you had in mind with this? After discussion with Hugh I removed the requirement that users not already be in/out of the group they are being added to/removed from.

File: auth-service/src/main/kotlin/uk/gov/communities/delta/auth/controllers/internal/AdminEditUserEmailController.kt [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+    private suspend fun updateEmail(call: ApplicationCall) {
+        val session = getSessionIfUserHasPermittedRole(
+            arrayOf(DeltaSystemRole.ADMIN), call
+        )
+        val requestData = call.receive<DeltaEmailChangeRequest>()
+        val requestedEmail = requestData.newEmail
+
+        val userToEdit = userLookupService.lookupUserByCn(requestData.userToEditCn)
+
+        logger.atInfo().log("Updating email and username for user {}", userToEdit.cn)
```
Comment (BenRamchandani) [inline auth-service/src/main/kotlin/uk/gov/communities/delta/auth/controllers/internal/AdminEditUserEmailController.kt:38]: Don't need this too

File: auth-service/src/main/kotlin/uk/gov/communities/delta/auth/controllers/internal/EditAccessGroupsController.kt (hunk 5 above)
Comment (BenRamchandani) [reply auth-service/src/main/kotlin/uk/gov/communities/delta/auth/controllers/internal/EditAccessGroupsController.kt]: Nothing specific, but we'll need to make sure we keep that behaviour

File: auth-service/src/test/kotlin/uk/gov/communities/delta/controllers/EditUserAccessGroupsControllerTest.kt [hunk 8]
```diff
@@ -235,13 +237,122 @@ class EditUserAccessGroupsControllerTest {
...
+            }
+            coVerify(exactly = 0) {
+                groupService.removeUserFromGroup(
+                    any(),
+                    any(),
+                    any(),
+                    any(),
+                    any()
+                )
+            }
```
Comment (BenRamchandani) [inline auth-service/src/test/kotlin/uk/gov/communities/delta/controllers/EditUserAccessGroupsControllerTest.kt:280]: Fine to keep it for consistency with other tests if you like, but the `confirmVerified` makes it unnecessary

**Review by BenRamchandani — APPROVED** [review]

**Pull request merged**